Refs #14831 — Added Template Style Guide

ryancheley opened the pull request:
Added more detail for the template style guide

ryancheley commented:
@adamchainz here's the PR based on our conversation from [this forum post](https://forum.djangoproject.com/t/django-template-style-guide-ticket-14831/28177/6)

I wasn't quite sure what you meant by so that isn't included in this

> Some more proposals:
> 
> - Top-level {% block %} tags for an {% extends %} template have no indentation.

nessita commented:
Hello @ryancheley, thank you for your contribution!

Please remember that when this work is ready for review, you need to adjust the ticket flags in the Trac system so this PR gets added to the "branches needing review" section of the [Django Developer Dahsboard](https://dashboard.djangoproject.com/). More information in [these docs](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-style).

ryancheley commented:
Hello @nessita! I have updated the flag of the trac ticket to ‘Has Patch: yes’ and updated to ‘Patch needs improvement’ since there’s an outstanding question. Thanks for pinging me on this!

felixxm commented on docs/internals/contributing/writing-code/coding-style.txt:
```diff
@@ -215,8 +215,37 @@ Imports
+
+  .. code-block:: html+django
+
+      {% extends %}
```
`extends` must extend something, so I'd add an ellipsis, or (even better) a template name: 
```suggestion
      {% extends "base.html" %}
```

felixxm commented on docs/internals/contributing/writing-code/coding-style.txt:
```diff
@@ -230,6 +259,74 @@ Template style
 
       {{foo}}
 
+  Put each ``{% load foo %}`` on it's own line
+
+  Do this:
+
+  .. code-block:: html+django
+
+      {% load foo %}
+      {% load bar %}
+
+  Don't do this:
+
+  .. code-block:: html+django
+
+      {% load foo bar %}
+
```
Is this really something that we want to recommend? There is nothing wrong in loading multiple libraries at once. I'd remove it.

felixxm replied:
And we do this in Django itself.

adamchainz replied:
Fair enough. I proposed the one-per-line rule to reduce diff noise, but it is a bit overbearing. Perhaps we could instead request the list be sorted alphabetically? That makes it a bit easier to skim.

felixxm commented on docs/internals/contributing/writing-code/coding-style.txt:
```diff
@@ -230,6 +259,74 @@ Template style
+
+      {% load foo bar %}
+
+  Put one (and only one) space between the curly brackets, 
+  the ``%`` and the tag contents.
```
Wrap new docs at 79 chars and use a serial comma.
```suggestion
  Put one (and only one) space between the curly brackets,  the ``%``, and the
  tag contents.
```

felixxm commented on docs/internals/contributing/writing-code/coding-style.txt:
```diff
@@ -215,8 +215,37 @@ Imports
-  brackets and the tag contents.
+* In Django template code: 
+
+  The ``{% extends %}`` should be the first non-comment line
```
We should render this as a list
```
* In Django template code, the ``{% extends %}`` should be ...

  ...

* Put one (and only one) space between the curly brackets and the tag contents.

  ....

* Put one (and only one) space between the curly brackets,  the ``%``, and the
  tag contents.

  ...
```
etc.

adamchainz replied:
Let’s also put the “In Django template code” in a full sentence before the list, something like:

```
Follow the below rules in Django template code.

* The ``{% extends %}`` should be ...
```

felixxm commented on docs/internals/contributing/writing-code/coding-style.txt:
```diff
@@ -230,6 +259,74 @@ Template style
+
+      {%foo%}
+
+  Each ``{% block %}`` should be named with the  ``{% endblock %}`` 
```
Avoid trailing spaces and multiple spaces:
```suggestion
  Each ``{% block %}`` should be named with the ``{% endblock %}``
```

felixxm commented:
@ryancheley Thanks :+1: I don't think we need any more recommendation than you proposed.

kezabelle commented on docs/internals/contributing/writing-code/coding-style.txt:
```diff
@@ -215,8 +215,37 @@ Imports
+
+  .. code-block:: html+django
+
+      # This is a comment
```
That's a _python_ comment, isn't it? It'd be `{# ... #}` (or, technically also `{% comment %} ... {% endcomment %}`) for the template language.

felixxm replied:
Good catch :+1: Please use `{# .... #}`.

adamchainz commented on docs/internals/contributing/writing-code/coding-style.txt:
```diff
@@ -215,8 +215,37 @@ Imports
+
+  .. code-block:: html+django
+
+      {% load foo %}
```
Let’s avoid using “foo” / “bar” since they’re meaningless and unrealistic examples that can confuse new people. I know the existing example did, but we can change that.

For example, here, we can name a real template library:

```suggestion
      {% load i18n %}
```

Below we could use `{% load humanize i18n %}`, and variables like `{{ user }}` , `{{ user.name|lower }}`, `{% block header %}`, etc.

adamchainz commented on docs/internals/contributing/writing-code/coding-style.txt:
```diff
@@ -230,6 +259,74 @@ Template style
+
+      {% endblock %}
+
+* Inside curly braces, tokens are separated by single spaces, 
+  except filters where the ``|`` has no spaces around it.
```
```suggestion
* Inside curly braces, separate tokens by single spaces, 
  except for around the ``.`` for attribute access and
  the ``|`` for a filter.
```

adamchainz commented on docs/internals/contributing/writing-code/coding-style.txt:
```diff
@@ -230,6 +259,74 @@ Template style
+* Inside curly braces, tokens are separated by single spaces, 
+  except filters where the ``|`` has no spaces around it.
+
+  Do this:
+
+  .. code-block:: html+django
+
+    {{ foo.bar|upper }}
+
+  Don't do this:
+
+  .. code-block:: html+django
+
+    {{ foo.bar | upper }}
```
```suggestion
  Do this:

  .. code-block:: html+django

    {% if user.name|lower == "admin" %}

  Don't do this:

  .. code-block:: html+django

    {% if user . name | lower  ==  "admin" %}
```

adamchainz commented on docs/internals/contributing/writing-code/coding-style.txt:
```diff
@@ -215,8 +215,37 @@ Imports
+
+      {% extends %}
+
+      Other code goes here
```
Let’s replace this with a realistic second line:

```suggestion
      {% block content %}
      ...
```

...and in below examples.

adamchainz reviewed (changes requested):
Thanks for this @ryancheley . I’ve added some comments that will hopefully make this easier to read.

>  Top-level {% block %} tags for an {% extends %} template have no indentation.

I meant something like this:

```
* Within a template using ``{% extends %}``, avoid indenting top-level
  ``{% block %}`` tags.

  Do this:

  .. code-block:: html+django

      {% extends "base.html" %}

      {% block content %}
      ...

  Don’t do this:

  .. code-block:: html+django

      {% extends "base.html" %}

          {% block content %}
          ...
```

felixxm commented:
@ryancheley Thanks :+1: 

I squashed commits and pushed small edits.

adamchainz commented on docs/internals/contributing/writing-code/coding-style.txt:
```diff
@@ -214,20 +214,150 @@ Imports
-  brackets and the tag contents.
+Follow the below rules in Django template code.
+
+* The ``{% extends %}`` should be the first non-comment line.
```
```suggestion
* ``{% extends %}`` should be the first non-comment line.
```

adamchainz commented on docs/internals/contributing/writing-code/coding-style.txt:
```diff
@@ -214,20 +214,150 @@ Imports
+        </h1>
+      {% endblock content %}
+
+* Put one (and only one) space between the curly brackets and the tag contents.
```
```suggestion
* Put exactly one space between ``{{``, variable contents, and ``}}``.
```

adamchainz commented on docs/internals/contributing/writing-code/coding-style.txt:
```diff
@@ -214,20 +214,150 @@ Imports
+
+      {{user}}
+
+* The ``{% load ... %}`` should list items in alphabetical order.
```
```suggestion
* In ``{% load ... %}``, list libraries in alphabetical order.
```

adamchainz commented on docs/internals/contributing/writing-code/coding-style.txt:
```diff
@@ -214,20 +214,150 @@ Imports
-      {{foo}}
+      {%load humanize%}
+
+* Each ``{% block %}`` should be named with the  ``{% endblock %}`` having the
+  matching name.
```
Using names on the same line is a bit verbose?

```suggestion
* Put the ``{% block %}`` tag name in the ``{% endblock %}`` tag if it is not on the same line.
```

felixxm replied:
I think it's still worth doing, as `block` are intended to be extended in templates.

adamchainz replied:
But the same syntax is used for filling in. `{% block title %}Something{% endblock %}` is clear enough to me.

felixxm replied:
Agreed, updated.

adamchainz commented on docs/internals/contributing/writing-code/coding-style.txt:
```diff
@@ -214,20 +214,150 @@ Imports
+
+      {% load l10 i18n tz %}
+
+* Put one (and only one) space between the curly brackets, the ``%``, and the
+  tag contents.
```
```suggestion
* Put exactly one space between ``{%``, tag contents, and ``%}``.
```

adamchainz reviewed (commented):
Some grammar improvements and suggestion for the block name rule.

carltongibson reviewed (commented):
@ryancheley @adamchainz @felixxm Q: Do you this the various _Don't do this_ blocks should be visually separated somehow? 

A big red ❌ would give folks a better chance of not reading it but still realising that this is a bad example they shouldn't copy. 

The only Sphinx tooling I can think of here is to put the _bad examples_ in an admonition. 

I don't know if that's worth it or not, so hence the _Q_.

felixxm commented on docs/internals/contributing/writing-code/coding-style.txt:
```diff
@@ -215,8 +215,51 @@ Imports
+
+In Django template code: 
+
+  The ``{% extends %}`` should be the first non-comment line
```
As far as I'm aware, I already fixed this :thinking:

adamchainz replied:
This was a comment I wrote 2 days ago that I accidentally didn't post until just now, mea culpa!

felixxm commented on docs/internals/contributing/writing-code/coding-style.txt:
```diff
@@ -215,8 +215,51 @@ Imports
+Follow the below rules in Django template code. 
+
+In Django template code: 
+
```
Should be already fixed.

felixxm commented:
> A big red ❌ would give folks a better chance of not reading it but still realising that this is a bad example they shouldn't copy.

We recently removed emojis from docs as they cause accessibility issues.

> The only Sphinx tooling I can think of here is to put the _bad examples_ in an admonition. **Update**: maybe adding a warning `:caption:` https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-option-code-block-caption ? 🤔

We use the same format `Do this`/`Don't do this` in other examples, so if we want do change it, we should do it consistently in all examples as a separate cleanup.
